### DISCUSSION TITLE
Create a second ceph cluster use Rook

### DISCUSSION THREAD

**Discussion by kencafe** [rook/rook#9902] [Q&A]
I want to create one more cluster using another node and its disks. 
I follow this guide https://github.com/rook/rook/blob/master/Documentation/ceph-advanced-configuration.md#deploying-a-second-cluster 

But I did not know what's the step I have to do after I create a new project: rook-ceph-secondary

I had created a new CRD cluster in project rook-ceph-secondary and specify a new cluster name but nothing happen after detect version.

The first cluster is running without any errors.

**Comment by travisn** [comment]
@kencafe 
- Is your intention to create an independent cluster? You will need to consider three nodes for each cluster, otherwise your reliability will suffer with a single node in the new cluster.
- Did you add nodeAffinity to each CephCluster CR so the clusters won't overlap? Otherwise, the original cluster might create OSDs on the new node already.
- Was an osd-prepare pod created on the new node? Its logs should show why OSDs were not created on the new node.

**Comment by kencafe** [comment]
@travisn This is my environment and my configurations:

I have six nodes: 03 worker nodes + 03 storage nodes
   - 03 worker nodes became members in the first cluster already: rook-ceph
   - 03 storage nodes available for the second cluster: rook-ceph-secondary

This is specifications in .yaml for the first cluster:

> 
> 
>   placement:
>     all:
>       nodeAffinity:
>         requiredDuringSchedulingIgnoredDuringExecution:
>           nodeSelectorTerms:
>             - matchExpressions:
>                 - key: storage
>                   operator: In
>                   values:
>                     - collocated               "label for worker node"
>   storage: 
>     useAllNodes: true
>     useAllDevices: false
>     deviceFilter: sdc
>     config:
> 
> 



This is specifications in .yaml for the second cluster:


>   placement:
>     all:
>       nodeAffinity:
>         requiredDuringSchedulingIgnoredDuringExecution:
>           nodeSelectorTerms:
>             - matchExpressions:
>                 - key: storage
>                   operator: In
>                   values:
>                     - separated               "label for storage node"
>   storage: 
>     useAllNodes: true
>     useAllDevices: false
>     deviceFilter: sdc
>     config:


For some key points you have mentioned:

- Is your intention to create an independent cluster? You will need to consider three nodes for each cluster, otherwise your reliability will suffer with a single node in the new cluster.

=> _I want to create secondary ceph cluster. I have 3 available nodes now._

- Did you add nodeAffinity to each CephCluster CR so the clusters won't overlap? Otherwise, the original cluster might create OSDs on the new node already.
=> _I have created node affinity in placement specification._ 

- Was an osd-prepare pod created on the new node? Its logs should show why OSDs were not created on the new node.

=> Only detect version pod is created and terminated immediately. 


All pods in the first cluster:

> csi-cephfsplugin-2gp8q                                3/3     Running     3          5d19h
> csi-cephfsplugin-6l5wc                                3/3     Running     3          5d19h
> csi-cephfsplugin-9stmb                                3/3     Running     3          5d19h
> csi-cephfsplugin-bx9p7                                3/3     Running     3          5d19h
> csi-cephfsplugin-gc5c5                                3/3     Running     3          5d19h
> csi-cephfsplugin-hvkmc                                3/3     Running     3          5d19h
> csi-cephfsplugin-lkmh9                                3/3     Running     3          5d19h
> csi-cephfsplugin-provisioner-6f54f6c477-9npzw         6/6     Running     0          5d12h
> csi-cephfsplugin-provisioner-6f54f6c477-zt8lx         6/6     Running     0          5d12h
> csi-cephfsplugin-tffb8                                3/3     Running     3          5d19h
> csi-cephfsplugin-vzjrl                                3/3     Running     3          5d19h
> csi-rbdplugin-4qf8c                                   3/3     Running     3          5d19h
> csi-rbdplugin-ctj49                                   3/3     Running     3          5d19h
> csi-rbdplugin-fwszn                                   3/3     Running     3          5d19h
> csi-rbdplugin-jk6hx                                   3/3     Running     3          5d19h
> csi-rbdplugin-mhzzv                                   3/3     Running     3          5d19h
> csi-rbdplugin-pdz6z                                   3/3     Running     3          5d19h
> csi-rbdplugin-provisioner-6d765b47d5-4m9wm            6/6     Running     0          5d12h
> csi-rbdplugin-provisioner-6d765b47d5-6dwjl            6/6     Running     0          5d12h
> csi-rbdplugin-wdsrd                                   3/3     Running     3          5d19h
> csi-rbdplugin-x9bct                                   3/3     Running     3          5d19h
> csi-rbdplugin-zpplh                                   3/3     Running     3          5d19h
> rook-ceph-crashcollector-storage01-6546985b65-89rpg   1/1     Running     0          5d12h
> rook-ceph-crashcollector-storage02-68c8c9cfb-k5sks    1/1     Running     0          5d12h
> rook-ceph-crashcollector-storage03-d45f984dc-hp2rm    1/1     Running     0          4d18h
> rook-ceph-crashcollector-worker01-cbfdbfd6d-xr5fc     1/1     Running     0          5d12h
> rook-ceph-crashcollector-worker02-5749cfc94c-9zq2d    1/1     Running     0          5d12h
> rook-ceph-crashcollector-worker03-5bd7b495fd-5zt82    1/1     Running     0          5d12h
> rook-ceph-mds-myfs-a-5875db9446-dr87f                 1/1     Running     0          5d12h
> rook-ceph-mds-myfs-b-6dbd4f8b5b-5l9cl                 1/1     Running     0          5d12h
> rook-ceph-mgr-a-7b754d5f8-dnq8p                       2/2     Running     0          5d12h
> rook-ceph-mgr-b-7486f7bff4-pnnjn                      2/2     Running     0          5d12h
> rook-ceph-mon-a-8695fdcf6d-zds8g                      1/1     Running     0          5d12h
> rook-ceph-mon-b-7f4548647b-qqtcj                      1/1     Running     0          5d12h
> rook-ceph-mon-c-7f7c9bc679-j8j6g                      1/1     Running     0          5d12h
> rook-ceph-operator-5df4d596d5-7jms9                   1/1     Running     0          5d12h
> rook-ceph-osd-0-86bcdf7499-6rzlx                      1/1     Running     0          5d12h
> rook-ceph-osd-1-595fd5c4b8-48tcm                      1/1     Running     0          5d12h
> rook-ceph-osd-2-7db594ccf-l78df                       1/1     Running     0          5d12h
> rook-ceph-osd-prepare-worker01--1-hqbcv               0/1     Completed   0          16h
> rook-ceph-osd-prepare-worker02--1-xnrds               0/1     Completed   0          16h
> rook-ceph-osd-prepare-worker03--1-l9p4m               0/1     Completed   0          16h
> rook-ceph-rgw-s3-store-a-6f648c75db-2s77z             1/1     Running     0          4d18h
> rook-ceph-tools-555c879675-4jth7                      1/1     Running     0          5d12h

**Comment by travisn** [comment]
>> Was an osd-prepare pod created on the new node? Its logs should show why OSDs were not created on the new node.

> => Only detect version pod is created and terminated immediately.

Can you share the operator log? If the only pod created in the second namespace is the detect version pod, the operator must have hit some error.

**Reply by kencafe** [reply]
This is operator logs:

> 2022-03-15 14:15:03.229233 I | op-mon: waiting for canary pod creation rook-ceph-mon-b-canary
> 2022-03-15 14:15:03.388528 I | op-mon: waiting for canary pod creation rook-ceph-mon-c-canary
> 2022-03-15 14:15:08.041014 I | op-mon: waiting for canary pod creation rook-ceph-mon-a-canary
> 2022-03-15 14:15:08.041172 E | op-mon: failed to schedule mon "a". failed to schedule canary pod(s)
> 2022-03-15 14:15:08.233649 I | op-mon: waiting for canary pod creation rook-ceph-mon-b-canary
> 2022-03-15 14:15:08.233762 E | op-mon: failed to schedule mon "b". failed to schedule canary pod(s)
> 2022-03-15 14:15:08.395582 I | op-mon: waiting for canary pod creation rook-ceph-mon-c-canary
> 2022-03-15 14:15:08.395627 E | op-mon: failed to schedule mon "c". failed to schedule canary pod(s)
> 2022-03-15 14:15:08.403699 I | op-mon: cleaning up canary monitor deployment "rook-ceph-mon-a-canary"
> 2022-03-15 14:15:08.417013 I | op-mon: cleaning up canary monitor deployment "rook-ceph-mon-b-canary"
> 2022-03-15 14:15:08.430069 I | op-mon: cleaning up canary monitor deployment "rook-ceph-mon-c-canary"
> 2022-03-15 14:15:08.464455 E | ceph-cluster-controller: failed to reconcile CephCluster "rook-ceph-secondary/rook-ceph-secondary". failed to reconcile cluster "rook-ceph-secondary": failed to configure local ceph cluster: failed to create cluster: failed to start ceph monitors: failed to assign pods to mons: failed to schedule mons
> 

It shows that failed to create mon pod. Maybe it conflicts name with the current running pods.


> rook-ceph-mon-a-8695fdcf6d-zds8g                      1/1     Running     0             6d
> rook-ceph-mon-b-7f4548647b-qqtcj                      1/1     Running     0             6d
> rook-ceph-mon-c-7f7c9bc679-j8j6g                      1/1     Running     0             6d

**Reply by travisn** [reply]
The mons for the new cluster were running in a different namespace from the first cluster, so the names shouldn't conflict. And since there is a node selector for a different set of nodes, nothing on the hosts should conflict with the mons. Did you see these canary pods created? I wonder if you could collect the kubectl description for them before they are deleted to see why they failed to schedule. Unfortunately, the operator didn't log anything more helpful for the cause.

**Comment by kencafe** [comment]
I did not see any pod in the new namespace.
One more thing, the detect version pod has appeared in the current running cluster: rook-ceph. 

I have specified namespace: rook-ceph-secondary in yaml:


> apiVersion: ceph.rook.io/v1
> kind: CephCluster
> metadata:
>   name: rook-ceph-secondary
>   namespace: rook-ceph-secondary
> spec:
>   cephVersion:
>     allowUnsupported: false
>   dataDirHostPath: /var/lib/rook
>   skipUpgradeChecks: false
>   continueUpgradeAfterChecksEvenIfNotHealthy: false
>   waitTimeoutForHealthyOSDInMinutes: 10
>   mon:
>     count: 3
>     allowMultiplePerNode: false
>   mgr:
>     count: 1
>     allowMultiplePerNode: false
>     modules:
>       - name: pg_autoscaler
>         enabled: true
>   dashboard:
>     enabled: true
>     ssl: true
>   monitoring:
>     enabled: false
>     rulesNamespace: rook-ceph-secondary
>   network:
>   crashCollector:
>     disable: false
>   cleanupPolicy:
>     confirmation: ""
>     sanitizeDisks:
>       method: quick
>       dataSource: zero
>       iteration: 1
>     allowUninstallWithVolumes: false
>   placement:
>     all:
>       nodeAffinity:
>         requiredDuringSchedulingIgnoredDuringExecution:
>           nodeSelectorTerms:
>             - matchExpressions:
>                 - key: storage
>                   operator: In
>                   values:
>                     - separated
>   annotations:
>   labels:
>   resources:
>   removeOSDsIfOutAndSafeToRemove: false
>   storage: 
>     useAllNodes: true
>     useAllDevices: false
>     deviceFilter: sdc
>     config:
>     onlyApplyOSDPlacement: false
>   disruptionManagement:
>     managePodBudgets: true
>     osdMaintenanceTimeout: 30
>     pgHealthCheckTimeout: 0
>     manageMachineDisruptionBudgets: false
>     machineDisruptionBudgetNamespace: openshift-machine-api
>   healthCheck:
>     daemonHealth:
>       mon:
>         disabled: false
>         interval: 45s
>       osd:
>         disabled: false
>         interval: 60s
>       status:
>         disabled: false
>         interval: 60s
>     livenessProbe:
>       mon:
>         disabled: false
>       mgr:
>         disabled: false
>       osd:
>         disabled: false
>     startupProbe:
>       mon:
>         disabled: false
>       mgr:
>         disabled: false
>       osd:
>         disabled: false

**Comment by travisn** [comment]
Do you see any resources in the new namespace such as configmaps or secrets created by rook? And it looks like the operator is creating the mon-canary deployments at least briefly even if they are failing to be scheduled, if you could watch for those. Restarting the operator can help trigger it again.

**Comment by kencafe** [comment]
I have restarted the operator pod and got logs. It shows that "existing maxMonID not found or failed to load. configmaps "rook-ceph-mon-endpoints" not found"

```
2022-03-15 15:18:58.861009 I | op-mon: start running mons
2022-03-15 15:18:58.929269 I | op-mon: creating mon secrets for a new cluster
2022-03-15 15:18:59.518440 I | op-mon: existing maxMonID not found or failed to load. configmaps "rook-ceph-mon-endpoints" not found
2022-03-15 15:18:59.723931 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph-secondary","monitors":[]}] data: mapping:{"node":{}} maxMonId:-1]
2022-03-15 15:19:00.518220 I | cephclient: writing config file /var/lib/rook/rook-ceph-secondary/rook-ceph-secondary.config
2022-03-15 15:19:00.518373 I | cephclient: generated admin config in /var/lib/rook/rook-ceph-secondary
2022-03-15 15:19:02.120107 I | op-mon: targeting the mon count 3
2022-03-15 15:19:02.128817 I | op-mon: created canary deployment rook-ceph-mon-a-canary
2022-03-15 15:19:02.144981 I | op-mon: created canary deployment rook-ceph-mon-b-canary
2022-03-15 15:19:02.156684 I | op-mon: created canary deployment rook-ceph-mon-c-canary
2022-03-15 15:19:02.320232 I | op-mon: waiting for canary pod creation rook-ceph-mon-a-canary
2022-03-15 15:19:02.519651 I | op-mon: waiting for canary pod creation rook-ceph-mon-b-canary
2022-03-15 15:19:02.717999 I | op-mon: waiting for canary pod creation rook-ceph-mon-c-canary
2022-03-15 15:19:04.935160 I | op-mon: parsing mon endpoints:
2022-03-15 15:19:04.935185 W | op-mon: ignoring invalid monitor
2022-03-15 15:19:04.935200 I | op-k8sutil: ROOK_OBC_WATCH_OPERATOR_NAMESPACE="true" (configmap)
2022-03-15 15:19:04.935207 I | op-bucket-prov: ceph bucket provisioner launched watching for provisioner "rook-ceph-secondary.ceph.rook.io/bucket"
2022-03-15 15:19:04.936404 I | op-bucket-prov: successfully reconciled bucket provisioner
I0315 15:19:04.936461       1 manager.go:135] objectbucket.io/provisioner-manager "msg"="starting provisioner"  "name"="rook-ceph-secondary.ceph.rook.io/bucket"
2022-03-15 15:19:07.323597 I | op-mon: waiting for canary pod creation rook-ceph-mon-a-canary
2022-03-15 15:19:07.524064 I | op-mon: waiting for canary pod creation rook-ceph-mon-b-canary
2022-03-15 15:19:07.720865 I | op-mon: waiting for canary pod creation rook-ceph-mon-c-canary
```

Then I check config maps and secrets and get a comparison between them:

```
[sysadmin@services examples]$ oc get configmaps -n rook-ceph

NAME                                DATA   AGE
kube-root-ca.crt                    1      6d8h
openshift-service-ca.crt            1      6d8h
rook-ceph-csi-config                1      6d8h
rook-ceph-csi-mapping-config        1      6d8h
rook-ceph-mon-endpoints             4      6d8h
rook-ceph-operator-config           17     6d8h
rook-ceph-pdbstatemap               2      6d8h
rook-ceph-rgw-s3-store-mime-types   1      5d7h
rook-config-override                1      6d8h


[sysadmin@services examples]$ oc get configmaps -n rook-ceph-secondary
NAME                       DATA   AGE
kube-root-ca.crt           1      3m4s
openshift-service-ca.crt   1      3m4s




[sysadmin@services examples]$ oc get secrets -n rook-ceph
NAME                                             TYPE                                  DATA   AGE
builder-dockercfg-z8k9d                          kubernetes.io/dockercfg               1      6d8h
builder-token-ktp94                              kubernetes.io/service-account-token   4      6d8h
builder-token-pzzzv                              kubernetes.io/service-account-token   4      6d8h
cluster-peer-token-rook-ceph                     kubernetes.io/rook                    2      6d8h
default-dockercfg-25clt                          kubernetes.io/dockercfg               1      6d8h
default-token-bs5cl                              kubernetes.io/service-account-token   4      6d8h
default-token-qqrg7                              kubernetes.io/service-account-token   4      6d8h
deployer-dockercfg-48nw2                         kubernetes.io/dockercfg               1      6d8h
deployer-token-6h8bx                             kubernetes.io/service-account-token   4      6d8h
deployer-token-fgl2t                             kubernetes.io/service-account-token   4      6d8h
rook-ceph-admin-keyring                          kubernetes.io/rook                    1      6d8h
rook-ceph-cmd-reporter-dockercfg-47p6b           kubernetes.io/dockercfg               1      6d8h
rook-ceph-cmd-reporter-token-4bz4m               kubernetes.io/service-account-token   4      6d8h
rook-ceph-cmd-reporter-token-pfqxt               kubernetes.io/service-account-token   4      6d8h
rook-ceph-config                                 kubernetes.io/rook                    2      6d8h
rook-ceph-crash-collector-keyring                kubernetes.io/rook                    1      6d8h
rook-ceph-dashboard-password                     kubernetes.io/rook                    1      6d8h
rook-ceph-mds-myfs-a-keyring                     kubernetes.io/rook                    1      6d8h
rook-ceph-mds-myfs-b-keyring                     kubernetes.io/rook                    1      6d8h
rook-ceph-mgr-a-keyring                          kubernetes.io/rook                    1      6d8h
rook-ceph-mgr-b-keyring                          kubernetes.io/rook                    1      6d8h
rook-ceph-mgr-dockercfg-7dprv                    kubernetes.io/dockercfg               1      6d8h
rook-ceph-mgr-token-8lv2k                        kubernetes.io/service-account-token   4      6d8h
rook-ceph-mgr-token-xkmqf                        kubernetes.io/service-account-token   4      6d8h
rook-ceph-mon                                    kubernetes.io/rook                    4      6d8h
rook-ceph-mons-keyring                           kubernetes.io/rook                    1      6d8h
rook-ceph-osd-dockercfg-9sfm5                    kubernetes.io/dockercfg               1      6d8h
rook-ceph-osd-token-87ck7                        kubernetes.io/service-account-token   4      6d8h
rook-ceph-osd-token-bl5rb                        kubernetes.io/service-account-token   4      6d8h
rook-ceph-purge-osd-dockercfg-bs2tb              kubernetes.io/dockercfg               1      6d8h
rook-ceph-purge-osd-token-9mdqx                  kubernetes.io/service-account-token   4      6d8h
rook-ceph-purge-osd-token-cc64v                  kubernetes.io/service-account-token   4      6d8h
rook-ceph-rgw-s3-store-a-keyring                 kubernetes.io/rook                    1      5d7h
rook-ceph-system-dockercfg-jl9km                 kubernetes.io/dockercfg               1      6d8h
rook-ceph-system-token-bmf2k                     kubernetes.io/service-account-token   4      6d8h
rook-ceph-system-token-sdnjm                     kubernetes.io/service-account-token   4      6d8h
rook-csi-cephfs-node                             kubernetes.io/rook                    2      6d8h
rook-csi-cephfs-plugin-sa-dockercfg-6hvlh        kubernetes.io/dockercfg               1      6d8h
rook-csi-cephfs-plugin-sa-token-4tz2t            kubernetes.io/service-account-token   4      6d8h
rook-csi-cephfs-plugin-sa-token-ch6nf            kubernetes.io/service-account-token   4      6d8h
rook-csi-cephfs-provisioner                      kubernetes.io/rook                    2      6d8h
rook-csi-cephfs-provisioner-sa-dockercfg-47kcp   kubernetes.io/dockercfg               1      6d8h
rook-csi-cephfs-provisioner-sa-token-qjx5s       kubernetes.io/service-account-token   4      6d8h
rook-csi-cephfs-provisioner-sa-token-r2xwl       kubernetes.io/service-account-token   4      6d8h
rook-csi-rbd-node                                kubernetes.io/rook                    2      6d8h
rook-csi-rbd-plugin-sa-dockercfg-xk2jq           kubernetes.io/dockercfg               1      6d8h
rook-csi-rbd-plugin-sa-token-grwhv               kubernetes.io/service-account-token   4      6d8h
rook-csi-rbd-plugin-sa-token-hbb85               kubernetes.io/service-account-token   4      6d8h
rook-csi-rbd-provisioner                         kubernetes.io/rook                    2      6d8h
rook-csi-rbd-provisioner-sa-dockercfg-9xlpr      kubernetes.io/dockercfg               1      6d8h
rook-csi-rbd-provisioner-sa-token-7p9fc          kubernetes.io/service-account-token   4      6d8h
rook-csi-rbd-provisioner-sa-token-hl55g          kubernetes.io/service-account-token   4      6d8h




[sysadmin@services examples]$ oc get secrets -n rook-ceph-secondary
NAME                                     TYPE                                  DATA   AGE
builder-dockercfg-h8d4p                  kubernetes.io/dockercfg               1      3m20s
builder-token-9wx4s                      kubernetes.io/service-account-token   4      3m20s
builder-token-txfjp                      kubernetes.io/service-account-token   4      3m20s
default-dockercfg-zwj8j                  kubernetes.io/dockercfg               1      3m20s
default-token-8mnj5                      kubernetes.io/service-account-token   4      3m20s
default-token-g9r77                      kubernetes.io/service-account-token   4      3m20s
deployer-dockercfg-74cdq                 kubernetes.io/dockercfg               1      3m20s
deployer-token-hhcs7                     kubernetes.io/service-account-token   4      3m20s
deployer-token-x5t76                     kubernetes.io/service-account-token   4      3m20s
rook-ceph-cmd-reporter-dockercfg-zdnfb   kubernetes.io/dockercfg               1      3m20s
rook-ceph-cmd-reporter-token-7kfsc       kubernetes.io/service-account-token   4      3m20s
rook-ceph-cmd-reporter-token-7kz5w       kubernetes.io/service-account-token   4      3m20s
rook-ceph-mgr-dockercfg-j52dj            kubernetes.io/dockercfg               1      3m20s
rook-ceph-mgr-token-2frmn                kubernetes.io/service-account-token   4      3m20s
rook-ceph-mgr-token-g2xxb                kubernetes.io/service-account-token   4      3m20s
rook-ceph-osd-dockercfg-8dlsk            kubernetes.io/dockercfg               1      3m20s
rook-ceph-osd-token-nrd5c                kubernetes.io/service-account-token   4      3m20s
rook-ceph-osd-token-pzvws                kubernetes.io/service-account-token   4      3m20s
rook-ceph-purge-osd-dockercfg-2vpwg      kubernetes.io/dockercfg               1      3m20s
rook-ceph-purge-osd-token-ljl8m          kubernetes.io/service-account-token   4      3m20s
rook-ceph-purge-osd-token-xg2s4          kubernetes.io/service-account-token   4      3m20s
```

**Comment by travisn** [comment]
This message indicates the `rook-ceph-mon-endpoints` configmap was created:
```
2022-03-15 15:18:59.723931 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph-secondary","monitors":[]}] data: mapping:{"node":{}} maxMonId:-1]
```
But yet I don't see that configmap in your list for the second namespace.

Were you able to catch a kubectl describe on the reason the mon-canary pods could not be created?

**Reply by kencafe** [reply]
Yes. I see the canary deployment could not create pods.

```
[sysadmin@services ~]$ oc get deployment
NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
rook-ceph-mon-a-canary   0/1     0            0           2m3s
rook-ceph-mon-b-canary   0/1     0            0           2m3s
rook-ceph-mon-c-canary   0/1     0            0           2m3s
[sysadmin@services ~]$ oc describe deployment rook-ceph-mon-a-canary
Name:               rook-ceph-mon-a-canary
Namespace:          rook-ceph-secondary
CreationTimestamp:  Tue, 15 Mar 2022 22:41:49 +0700
Labels:             app=rook-ceph-mon
                    app.kubernetes.io/component=cephclusters.ceph.rook.io
                    app.kubernetes.io/created-by=rook-ceph-operator
                    app.kubernetes.io/instance=a
                    app.kubernetes.io/managed-by=rook-ceph-operator
                    app.kubernetes.io/name=ceph-mon
                    app.kubernetes.io/part-of=rook-ceph-secondary
                    ceph-version=16.2.7-0
                    ceph_daemon_id=a
                    ceph_daemon_type=mon
                    mon=a
                    mon_canary=true
                    mon_cluster=rook-ceph-secondary
                    rook-version=v1.8.0-alpha.0.383.g7ef5a0462
                    rook.io/operator-namespace=rook-ceph
                    rook_cluster=rook-ceph-secondary
Annotations:        deployment.kubernetes.io/revision: 1
Selector:           app=rook-ceph-mon,ceph_daemon_id=a,mon=a,mon_canary=true,mon_cluster=rook-ceph-secondary,rook_cluster=rook-ceph-secondary
Replicas:           1 desired | 0 updated | 0 total | 0 available | 1 unavailable
StrategyType:       Recreate
MinReadySeconds:    0
Pod Template:
  Labels:  app=rook-ceph-mon
           app.kubernetes.io/component=cephclusters.ceph.rook.io
           app.kubernetes.io/created-by=rook-ceph-operator
           app.kubernetes.io/instance=a
           app.kubernetes.io/managed-by=rook-ceph-operator
           app.kubernetes.io/name=ceph-mon
           app.kubernetes.io/part-of=rook-ceph-secondary
           ceph_daemon_id=a
           ceph_daemon_type=mon
           mon=a
           mon_canary=true
           mon_cluster=rook-ceph-secondary
           rook.io/operator-namespace=rook-ceph
           rook_cluster=rook-ceph-secondary
  Containers:
   mon:
    Image:      rook/ceph:master
    Port:       6789/TCP
    Host Port:  0/TCP
    Command:
      sleep
    Args:
      3600
    Startup:  exec [env -i sh -c ceph --admin-daemon /run/ceph/ceph-mon.a.asok mon_status] delay=10s timeout=1s period=10s #success=1 #failure=6
    Environment:
      CONTAINER_IMAGE:                quay.io/ceph/ceph:v16.2.7
      POD_NAME:                        (v1:metadata.name)
      POD_NAMESPACE:                   (v1:metadata.namespace)
      NODE_NAME:                       (v1:spec.nodeName)
      POD_MEMORY_LIMIT:               node allocatable (limits.memory)
      POD_MEMORY_REQUEST:             0 (requests.memory)
      POD_CPU_LIMIT:                  node allocatable (limits.cpu)
      POD_CPU_REQUEST:                0 (requests.cpu)
      ROOK_CEPH_MON_HOST:             <set to the key 'mon_host' in secret 'rook-ceph-config'>             Optional: false
      ROOK_CEPH_MON_INITIAL_MEMBERS:  <set to the key 'mon_initial_members' in secret 'rook-ceph-config'>  Optional: false
      ROOK_POD_IP:                     (v1:status.podIP)
    Mounts:
      /etc/ceph from rook-config-override (ro)
      /etc/ceph/keyring-store/ from rook-ceph-mons-keyring (ro)
      /var/lib/ceph/crash from rook-ceph-crash (rw)
      /var/lib/ceph/mon/ceph-a from ceph-daemon-data (rw)
      /var/log/ceph from rook-ceph-log (rw)
  Volumes:
   rook-config-override:
    Type:               Projected (a volume that contains injected data from multiple sources)
    ConfigMapName:      rook-config-override
    ConfigMapOptional:  <nil>
   rook-ceph-mons-keyring:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  rook-ceph-mons-keyring
    Optional:    false
   rook-ceph-log:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/rook/rook-ceph-secondary/log
    HostPathType:
   rook-ceph-crash:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/rook/rook-ceph-secondary/crash
    HostPathType:
   ceph-daemon-data:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/rook/mon-a/data
    HostPathType:
Conditions:
  Type             Status  Reason
  ----             ------  ------
  Progressing      True    NewReplicaSetCreated
  Available        False   MinimumReplicasUnavailable
  ReplicaFailure   True    FailedCreate
OldReplicaSets:    <none>
NewReplicaSet:     rook-ceph-mon-a-canary-584586877c (0/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  2m9s  deployment-controller  Scaled up replica set rook-ceph-mon-a-canary-584586877c to 1
```

**Reply by kencafe** [reply]
Last canary deployment status showed error related scc:

```
status:
  conditions:
  - lastTransitionTime: "2022-03-15T16:12:54Z"
    lastUpdateTime: "2022-03-15T16:12:54Z"
    message: Created new replica set "rook-ceph-mon-a-canary-584586877c"
    reason: NewReplicaSetCreated
    status: "True"
    type: Progressing
  - lastTransitionTime: "2022-03-15T16:12:54Z"
    lastUpdateTime: "2022-03-15T16:12:54Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2022-03-15T16:12:54Z"
    lastUpdateTime: "2022-03-15T16:12:54Z"
    message: 'pods "rook-ceph-mon-a-canary-584586877c-" is forbidden: unable to validate
      against any security context constraint: [provider "rook-ceph": Forbidden: not
      usable by user or serviceaccount, provider "rook-ceph-csi": Forbidden: not usable
      by user or serviceaccount, provider "allow-nexus-userid-200": Forbidden: not
      usable by user or serviceaccount, provider "nonroot": Forbidden: not usable
      by user or serviceaccount, provider "anyuid": Forbidden: not usable by user
      or serviceaccount, spec.containers[0].securityContext.privileged: Invalid value:
      true: Privileged containers are not allowed, provider "hostmount-anyuid": Forbidden:
      not usable by user or serviceaccount, provider "log-collector-scc": Forbidden:
      not usable by user or serviceaccount, provider "machine-api-termination-handler":
      Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden:
      not usable by user or serviceaccount, provider "hostaccess": Forbidden: not
      usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable
      by user or serviceaccount, provider "privileged": Forbidden: not usable by user
      or serviceaccount]'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```

**Reply by travisn** [reply]
Aha, this is openshift? Try adding the new namespace for each of these lines in the scc: https://github.com/rook/rook/blob/master/deploy/examples/operator-openshift.yaml#L46-L49

**Reply by kencafe** [reply]
Hi @travisn 

Thanks for your time to help me find the solution.
Finally, I need to add SCC to the serviceaccount after the new namespace is created.

'''
If other namespaces or service accounts are configured, they need to be updated here.
  - system:serviceaccount:rook-ceph-secondary:rook-ceph-system # serviceaccount:namespace:operator
  - system:serviceaccount:rook-ceph-secondary:default # serviceaccount:namespace:cluster
  - system:serviceaccount:rook-ceph-secondary:rook-ceph-mgr # serviceaccount:namespace:cluster
  - system:serviceaccount:rook-ceph-secondary:rook-ceph-osd # serviceaccount:namespace:cluster

'''

Many thanks again.

**Comment by BlaineEXE** [comment]
@kencafe please use triple-grave marks (```)to post commands and output to better preserve output, not block quotes. I have edited a few of your comments already if you want to look and see.

**Reply by kencafe** [reply]
@BlaineEXE 

Thank you for your notice. I will use it next time. :)

**Comment by karr1gan** [comment]
Following the discussion I wonder if I am able to create a second cluster on the same nodes but different disks and in different namespace. I've successfully deployed cluster on the /dev/sdb rook-ceph ns but i want to test disaster recovery so i want to deploy cluster on /dev/sdc and rook-ceph-2 ns. Is it feasible?

```
Number  Start (sector)    End (sector)  Size       Code  Name
2025-12-31 10:42:13.526826 I | cephosd: setting device type "hdd" for device "/dev/sdc"
2025-12-31 10:42:13.526832 I | cephosd: setting device class "hdd" for device "/dev/sdc"
2025-12-31 10:42:13.526840 I | cephosd: skipping osd.0: "AASDAD-2330-4c53-ADSAD-ADSADSA" belonging to a different ceph cluster "8b7b8f38-8e2b-40a6-b50d-d4c64c228dc1"
2025-12-31 10:42:13.526843 I | cephosd: 1 ceph-volume raw osd devices configured on this node
2025-12-31 10:42:13.526872 I | cephosd: devices = [{ID:0 Cluster:ceph UUID:cASDSAD3c-c276-43d0-81d4-91ASD6dADASf DevicePartUUID: DeviceClass:hdd BlockPath:/dev/sdc MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=node1 LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName: DeviceType:hdd CephxStatus:{KeyGeneration:0 KeyCephVersiosudo rm -rf /var/lib/rook/mon-d/n:}}]
2025-12-31 10:42:13.529968 E | op-osd: failed to set node or PVC "node1" status to "completed" for osd orchestration.
```

I've deployed it third time because previously some leftovers blocked proper deployment of the second cluster.

During the clean up process I've made following steps:
```
sudo umount /dev/sdc* 2>/devps aux | grep ceph | grep -v grep/null || true
sudo wipefs -a /dev/sdc
sudo modprobe -r rbd
sudo modprobe -r libceph
ls -lah /var/lib/rook
sudo rm -rf /var/lib/rook/exporter
sudo rm -rf /var/lib/rook/mon-a/
sudo rm -rf /var/lib/rook/mon-b/
sudo rm -rf /var/lib/rook/mon-d/
sudo rm -rf /var/lib/rook/rook-ceph/
sudo rm -rf /var/lib/rook/rook-ceph-2/

umount and delete HDD drive from the host and install new one without formatting.
```

Maybe some leftovers still remained so they potentially blocked correct deployment of the second cluster. Am I missing something during cleanup process, maybe additional catalogues should be deleteted. Or maybe it's not possible?

Or I should deploy second Rook Operator with somehow nameOverride so the resources dont overlap and conflict with each other. Unfortunately, I couldn't find the solution how to override the name of the Operator and therefore Helm conflicts occured.

**Reply by travisn** [reply]
Yes, this is possible to configure two CephCluster CRs in the same K8s cluster. A few requirements for each CephCluster CR:
1. Must be created in a separate K8s namespace
2. Configured with a deviceFilter or specific devices so they won't overlap when creating OSDs
3. Different `dataDirHostPath` so the mons will persist their data under different host paths

[This topic](https://rook.io/docs/rook/latest-release/Storage-Configuration/Advanced/ceph-configuration/#deploying-a-second-cluster) may help. 

If using the helm charts, you would create one `rook-ceph` chart for the operator, then multiple `rook-ceph-cluster` charts in separate namespaces. The first cluster could be in the same namespace as the operator, then other clusters would need to configure the [operatorNamespace](https://github.com/rook/rook/blob/master/deploy/charts/rook-ceph-cluster/values.yaml#L6) override.

**Reply by karr1gan** [reply]
Thank you @travisn, working like a charm :+1:
